File: typing_speed.js
#!/usr/bin/env node
// typing_speed.js
const { program } = require('commander');
const fs = require('fs');
const chalk = require('chalk');
const readline = require('readline');

// Словари
const WORDS_EN = [
    "apple", "orange", "banana", "grape", "peach", "mango", "lemon", "melon", "berry", "plum",
    "cloud", "storm", "rain", "snow", "wind", "sun", "moon", "star", "sky", "tree",
    "book", "table", "chair", "window", "door", "floor", "wall", "roof", "garden", "flower",
    "happy", "sad", "angry", "calm", "brave", "smart", "kind", "funny", "serious", "wild",
    "quick", "brown", "fox", "jumps", "lazy", "dog", "hello", "world", "python", "code"
];
const WORDS_RU = [
    "яблоко", "апельсин", "банан", "виноград", "персик", "манго", "лимон", "дыня", "ягода", "слива",
    "облако", "буря", "дождь", "снег", "ветер", "солнце", "луна", "звезда", "небо", "дерево",
    "книга", "стол", "стул", "окно", "дверь", "пол", "стена", "крыша", "сад", "цветок",
    "счастливый", "грустный", "злой", "спокойный", "храбрый", "умный", "добрый", "смешной", "серьёзный", "дикий",
    "быстрый", "коричневый", "лиса", "прыгает", "ленивый", "собака", "привет", "мир", "питон", "код"
];

class TypingSpeed {
    constructor(language = 'en', mode = 'words', length = 30, exportStats = null) {
        this.language = language;
        this.mode = mode;
        this.length = length;
        this.exportStats = exportStats;
        this.words = language === 'en' ? WORDS_EN : WORDS_RU;
        this.text = this._generateText();
        this.startTime = null;
        this.endTime = null;
        this.errors = 0;
        this.totalChars = 0;
        this.historyFile = 'typing_history.json';
        this.history = this._loadHistory();
        this.rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
        });
    }

    _generateText() {
        if (this.mode === 'chars') {
            let chars = this.words.join('').slice(0, this.length);
            return chars.split('').slice(0, this.length).join('');
        } else if (this.mode === 'sentences') {
            const shuffled = this.words.sort(() => Math.random() - 0.5);
            const selected = shuffled.slice(0, Math.min(this.length, shuffled.length));
            return selected.join(' ').charAt(0).toUpperCase() + selected.join(' ').slice(1) + '.';
        } else {
            const shuffled = this.words.sort(() => Math.random() - 0.5);
            const selected = shuffled.slice(0, Math.min(this.length, shuffled.length));
            return selected.join(' ');
        }
    }

    _loadHistory() {
        try {
            if (fs.existsSync(this.historyFile)) {
                return JSON.parse(fs.readFileSync(this.historyFile, 'utf8'));
            }
        } catch (e) {}
        return [];
    }

    _saveHistory(stats) {
        this.history.push(stats);
        fs.writeFileSync(this.historyFile, JSON.stringify(this.history, null, 2));
    }

    _question(prompt) {
        return new Promise((resolve) => {
            this.rl.question(prompt, resolve);
        });
    }

    async run() {
        console.log(chalk.cyan('Введите следующий текст как можно быстрее и точнее:'));
        console.log(chalk.white(this.text));
        console.log('\nНажмите Enter, чтобы начать...');
        await this._question('');
        this.startTime = Date.now();
        console.log(chalk.green('> '));
        const userInput = await this._question('');
        this.endTime = Date.now();
        this.totalChars = userInput.length;
        const minLen = Math.min(this.text.length, userInput.length);
        this.errors = 0;
        for (let i = 0; i < minLen; i++) {
            if (this.text[i] !== userInput[i]) this.errors++;
        }
        if (userInput.length > this.text.length) this.errors += userInput.length - this.text.length;
        const elapsed = (this.endTime - this.startTime) / 1000;
        const cpm = this.totalChars / elapsed * 60;
        const wpm = cpm / 5;
        const accuracy = this.totalChars > 0 ? ((this.totalChars - this.errors) / this.totalChars * 100) : 0;
        const stats = {
            timestamp: new Date().toISOString(),
            language: this.language,
            mode: this.mode,
            length: this.length,
            text_len: this.text.length,
            typed_len: this.totalChars,
            errors: this.errors,
            cpm: parseFloat(cpm.toFixed(2)),
            wpm: parseFloat(wpm.toFixed(2)),
            accuracy: parseFloat(accuracy.toFixed(2)),
            time_sec: parseFloat(elapsed.toFixed(2))
        };
        console.log(chalk.cyan('\n--- Результаты ---'));
        console.log(chalk.green(`Скорость: ${stats.cpm.toFixed(1)} симв/мин (${stats.wpm.toFixed(1)} слов/мин)`));
        console.log(chalk.yellow(`Точность: ${stats.accuracy.toFixed(1)}%`));
        console.log(chalk.magenta(`Время: ${stats.time_sec.toFixed(2)} сек`));
        console.log(chalk.red(`Ошибок: ${stats.errors} (из ${stats.typed_len})`));
        this._saveHistory(stats);
        if (this.exportStats) {
            this._exportStats(stats);
        }
        this.rl.close();
        return stats;
    }

    _exportStats(stats) {
        const filename = this.exportStats;
        const ext = filename.split('.').pop().toLowerCase();
        if (ext === 'json') {
            fs.writeFileSync(filename, JSON.stringify(stats, null, 2));
        } else if (ext === 'csv') {
            const header = Object.keys(stats).join(',') + '\n';
            const row = Object.values(stats).join(',') + '\n';
            fs.writeFileSync(filename, header + row);
        }
        console.log(chalk.green(`Статистика сохранена в ${filename}`));
    }

    showHistory() {
        if (this.history.length === 0) {
            console.log(chalk.yellow('История пуста.'));
            return;
        }
        console.log(chalk.cyan('📊 История тренировок:'));
        for (let i = 0; i < Math.min(10, this.history.length); i++) {
            const entry = this.history[this.history.length - 1 - i];
            console.log(`  ${i+1}. ${entry.timestamp.slice(0,19)} | ${entry.language} | ${entry.mode} | CPM: ${Math.round(entry.cpm)} | WPM: ${Math.round(entry.wpm)} | Точность: ${entry.accuracy.toFixed(1)}%`);
        }
    }
}

program
    .option('-l, --language <lang>', 'Язык: en, ru', 'en')
    .option('-m, --mode <mode>', 'Режим: words, sentences, chars', 'words')
    .option('-L, --length <number>', 'Количество слов или символов', parseInt, 30)
    .option('-e, --export-stats <file>', 'Экспорт статистики')
    .option('--history', 'Показать историю')
    .parse(process.argv);

const opts = program.opts();

const trainer = new TypingSpeed(opts.language, opts.mode, opts.length, opts.exportStats);

if (opts.history) {
    trainer.showHistory();
} else {
    trainer.run();
}

File: TypingSpeed.cs
// TypingSpeed.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace TypingSpeed
{
    class Program
    {
        static void Main(string[] args)
        {
            var opts = ParseArgs(args);
            var trainer = new TypingSpeed(opts);
            if (opts.History) trainer.ShowHistory();
            else trainer.Run();
        }

        static Options ParseArgs(string[] args)
        {
            var opts = new Options();
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--language": opts.Language = args[++i]; break;
                    case "--mode": opts.Mode = args[++i]; break;
                    case "--length": opts.Length = int.Parse(args[++i]); break;
                    case "--export-stats": opts.ExportStats = args[++i]; break;
                    case "--history": opts.History = true; break;
                }
            }
            return opts;
        }

        class Options
        {
            public string Language { get; set; } = "en";
            public string Mode { get; set; } = "words";
            public int Length { get; set; } = 30;
            public string ExportStats { get; set; }
            public bool History { get; set; }
        }

        class Stats
        {
            public string Timestamp { get; set; }
            public string Language { get; set; }
            public string Mode { get; set; }
            public int Length { get; set; }
            public int TextLen { get; set; }
            public int TypedLen { get; set; }
            public int Errors { get; set; }
            public double Cpm { get; set; }
            public double Wpm { get; set; }
            public double Accuracy { get; set; }
            public double TimeSec { get; set; }
        }

        class TypingSpeed
        {
            private readonly Options opts;
            private readonly string[] wordsEn = {
                "apple","orange","banana","grape","peach","mango","lemon","melon","berry","plum",
                "cloud","storm","rain","snow","wind","sun","moon","star","sky","tree",
                "book","table","chair","window","door","floor","wall","roof","garden","flower",
                "happy","sad","angry","calm","brave","smart","kind","funny","serious","wild",
                "quick","brown","fox","jumps","lazy","dog","hello","world","python","code"
            };
            private readonly string[] wordsRu = {
                "яблоко","апельсин","банан","виноград","персик","манго","лимон","дыня","ягода","слива",
                "облако","буря","дождь","снег","ветер","солнце","луна","звезда","небо","дерево",
                "книга","стол","стул","окно","дверь","пол","стена","крыша","сад","цветок",
                "счастливый","грустный","злой","спокойный","храбрый","умный","добрый","смешной","серьёзный","дикий",
                "быстрый","коричневый","лиса","прыгает","ленивый","собака","привет","мир","питон","код"
            };
            private string text;
            private DateTime startTime;
            private DateTime endTime;
            private int errors;
            private int totalChars;
            private readonly string historyFile = "typing_history.json";
            private List<Stats> history = new List<Stats>();

            public TypingSpeed(Options opts)
            {
                this.opts = opts;
                text = GenerateText();
                LoadHistory();
            }

            private string GenerateText()
            {
                var words = opts.Language == "ru" ? wordsRu : wordsEn;
                var rnd = new Random();
                if (opts.Mode == "chars")
                {
                    var all = string.Join("", words);
                    var chars = all.ToCharArray();
                    var sb = new System.Text.StringBuilder();
                    for (int i = 0; i < Math.Min(opts.Length, chars.Length); i++)
                        sb.Append(chars[rnd.Next(chars.Length)]);
                    return sb.ToString();
                }
                else if (opts.Mode == "sentences")
                {
                    var shuffled = words.OrderBy(_ => rnd.Next()).ToList();
                    var selected = shuffled.Take(Math.Min(opts.Length, shuffled.Count)).ToList();
                    var sentence = string.Join(" ", selected);
                    if (sentence.Length > 0)
                        sentence = char.ToUpper(sentence[0]) + sentence.Substring(1) + ".";
                    return sentence;
                }
                else
                {
                    var shuffled = words.OrderBy(_ => rnd.Next()).ToList();
                    var selected = shuffled.Take(Math.Min(opts.Length, shuffled.Count));
                    return string.Join(" ", selected);
                }
            }

            private void LoadHistory()
            {
                try
                {
                    if (File.Exists(historyFile))
                    {
                        string json = File.ReadAllText(historyFile);
                        history = JsonSerializer.Deserialize<List<Stats>>(json) ?? new List<Stats>();
                    }
                }
                catch { history = new List<Stats>(); }
            }

            private void SaveHistory(Stats stats)
            {
                history.Add(stats);
                string json = JsonSerializer.Serialize(history, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(historyFile, json);
            }

            public void Run()
            {
                Console.ForegroundColor = ConsoleColor.Cyan;
                Console.WriteLine("Введите следующий текст как можно быстрее и точнее:");
                Console.ForegroundColor = ConsoleColor.White;
                Console.WriteLine(text);
                Console.ResetColor();
                Console.WriteLine("\nНажмите Enter, чтобы начать...");
                Console.ReadLine();
                Console.ForegroundColor = ConsoleColor.Green;
                Console.Write("> ");
                Console.ResetColor();
                startTime = DateTime.UtcNow;
                string input = Console.ReadLine() ?? "";
                endTime = DateTime.UtcNow;
                totalChars = input.Length;
                int minLen = Math.Min(text.Length, input.Length);
                errors = 0;
                for (int i = 0; i < minLen; i++)
                    if (text[i] != input[i]) errors++;
                if (input.Length > text.Length) errors += input.Length - text.Length;
                double elapsed = (endTime - startTime).TotalSeconds;
                double cpm = totalChars / elapsed * 60;
                double wpm = cpm / 5;
                double accuracy = totalChars > 0 ? (double)(totalChars - errors) / totalChars * 100 : 0;
                var stats = new Stats
                {
                    Timestamp = DateTime.UtcNow.ToString("o"),
                    Language = opts.Language,
                    Mode = opts.Mode,
                    Length = opts.Length,
                    TextLen = text.Length,
                    TypedLen = totalChars,
                    Errors = errors,
                    Cpm = Math.Round(cpm, 2),
                    Wpm = Math.Round(wpm, 2),
                    Accuracy = Math.Round(accuracy, 2),
                    TimeSec = Math.Round(elapsed, 2)
                };
                Console.ForegroundColor = ConsoleColor.Cyan;
                Console.WriteLine("\n--- Результаты ---");
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"Скорость: {stats.Cpm:F1} симв/мин ({stats.Wpm:F1} слов/мин)");
                Console.ForegroundColor = ConsoleColor.Yellow;
                Console.WriteLine($"Точность: {stats.Accuracy:F1}%");
                Console.ForegroundColor = ConsoleColor.Magenta;
                Console.WriteLine($"Время: {stats.TimeSec:F2} сек");
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine($"Ошибок: {stats.Errors} (из {stats.TypedLen})");
                Console.ResetColor();
                SaveHistory(stats);
                if (!string.IsNullOrEmpty(opts.ExportStats))
                    ExportStats(stats);
            }

            private void ExportStats(Stats stats)
            {
                string ext = Path.GetExtension(opts.ExportStats).ToLower().TrimStart('.');
                string content;
                if (ext == "json")
                {
                    content = JsonSerializer.Serialize(stats, new JsonSerializerOptions { WriteIndented = true });
                }
                else if (ext == "csv")
                {
                    content = "timestamp,language,mode,length,text_len,typed_len,errors,cpm,wpm,accuracy,time_sec\n";
                    content += $"{stats.Timestamp},{stats.Language},{stats.Mode},{stats.Length},{stats.TextLen},{stats.TypedLen},{stats.Errors},{stats.Cpm},{stats.Wpm},{stats.Accuracy},{stats.TimeSec}\n";
                }
                else
                {
                    content = stats.ToString();
                }
                File.WriteAllText(opts.ExportStats, content);
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"Статистика сохранена в {opts.ExportStats}");
                Console.ResetColor();
            }

            public void ShowHistory()
            {
                if (history.Count == 0)
                {
                    Console.ForegroundColor = ConsoleColor.Yellow;
                    Console.WriteLine("История пуста.");
                    Console.ResetColor();
                    return;
                }
                Console.ForegroundColor = ConsoleColor.Cyan;
                Console.WriteLine("📊 История тренировок:");
                Console.ResetColor();
                int start = Math.Max(0, history.Count - 10);
                for (int i = history.Count - 1; i >= start; i--)
                {
                    var s = history[i];
                    Console.WriteLine($"  {history.Count - i}. {s.Timestamp.Substring(0,19).Replace('T',' ')} | {s.Language} | {s.Mode} | CPM: {Math.Round(s.Cpm)} | WPM: {Math.Round(s.Wpm)} | Точность: {s.Accuracy:F1}%");
                }
            }
        }
    }
}
